src/utils/layout-helpers.ts

```diff
@@ -0,0 +1,374 @@
+  const pack = getPackageJSON(root);
+  const hasIndexJs = existsSync(join(root, 'index.js'));
+  return isEmberAddon(pack) && hasIndexJs;
+}
+
+export function getProjectInRepoAddonsRoots(root: string) {
+  const prefix = isMuApp(root) ? 'packages' : 'lib';
+  const addons = safeWalkSync(
+    join(root, prefix),
+    {
+      directories: true,
+      globs: ['**/package.json']
+    }
+  );
+  const roots: string[] = [];
+  addons.map((relativePath: string) => {
+    return dirname(join(root, prefix, relativePath));
+  })
+  .filter((packageRoot: string) => isProjectAddonRoot(packageRoot))
+  .forEach((validRoot: string) => {
+    roots.push(validRoot);
+    getProjectAddonsRoots(validRoot, roots).forEach((relatedRoot: string) => {
+      if (!roots.includes(relatedRoot)) {
+        roots.push(relatedRoot);
+      }
+    });
+  });
+  return roots;
+}
+
+export function getProjectAddonsRoots(root: string, resolvedItems: string[] = [], packageFolderName = 'node_modules') {
+  // log('getProjectAddonsInfo', root);
+  const pack = getPackageJSON(root);
+  if (resolvedItems.length) {
+    if (!isEmberAddon(pack)) {
+      return [];
+    }
+  }
+  // log('getPackageJSON', pack);
+  const items = resolvedItems.length ?
+    [
+      ...Object.keys(pack.dependencies || {}),
+      ...Object.keys(pack.peerDependencies || {})
+    ] : [
+    ...Object.keys(pack.dependencies || {}),
+    ...Object.keys(pack.peerDependencies || {}),
+    ...Object.keys(pack.devDependencies || {})
+  ];
+  // log('items', items);
+
+  const roots = items
+    .map((item: string) => {
+      return resolvePackageRoot(root, item, packageFolderName);
+    })
+    .filter((p: string | boolean) => {
+      return p !== false;
+    });
+  const recursiveRoots: string[] = resolvedItems.slice(0);
+  roots.forEach((rootItem: string) => {
+    if (!recursiveRoots.includes(rootItem)) {
+      recursiveRoots.push(rootItem);
+      getProjectAddonsRoots(rootItem, recursiveRoots, packageFolderName).forEach((item: string) => {
+        if (!recursiveRoots.includes(item)) {
+          recursiveRoots.push(item);
+        }
+      });
+    }
+  });
+  return recursiveRoots;
+}
+
+export function getPackageJSON(file: string) {
+  try {
+    const result = JSON.parse(readFileSync(join(file, 'package.json'), 'utf8'));
+    return result;
+  } catch (e) {
+    return {};
+  }
+}
+
+export function isEmberAddon(info: any) {
+  return info.keywords && info.keywords.includes('ember-addon');
+}
+
+function addonVersion(info: any) {
+  if (!isEmberAddon(info)) {
+    return null;
+  }
+  return isEmberAddonV2(info) ? 2 : 1;
+}
+
+function isEmberAddonV2(info: any) {
+  return info['ember-addon'] && info['ember-addon'].version === 2;
+}
+
+export function isTemplatePath(filePath: string) {
+  return filePath.endsWith('.hbs');
+}
+
+export function hasAddonFolderInPath(name: string) {
+  return name.includes(sep + 'addon' + sep);
+}
+
+export function getProjectAddonsInfo(root: string) {
```

Comment:
We can improve the type experience throughout the addon by eliminating the use of `any[]` in the body of this in favor of `CompletionItem[]`. We can also improve the performance of the implementation by (a) not allocating the fresh arrays when pushing into `extractedData`, and (b) not iterating through the whole set of generated arrays *again* after building it. That can also give us fewer and clearer lines of code. (See inbound commit to this effect!)